lib/diaspora/federation/entities.rb

```diff
@@ -200,37 +177,29 @@ def self.reshare(reshare)
         )
       end
 
-      def self.retraction(target)
+      def self.retraction(retraction)
+        case retraction.data[:target_type]
+        when "Contact"
```

Comment:
> It's pretty much the same thing as what you do, but it uses inheritance instead of branching which is a bit cleaner IMO.

Yes, looks good to me, I took your diff and added it as it's own commit (I wasn't able to add it as fixup for any commit, because all other changes around `Retraction`s are in different commits, so it would have had to many conflicts) and used you as author (since you did the work for that, I hope that's OK for you).

I changed only one little thing: I moved the `retraction_class` as second parameter, so `opts` is the last parameter again. Please tell me what you think about it. (We could let it as the last parameter and set a default value to stay backward compatible with jobs still in the queue, but since this is a major release we can do the "run sidekiq for extra 5 minutes to process queued jobs"-migration again, so I don't think we need to stay compatible with old jobs.)

I also added some tests for it.

> Actually this responsibilities aren't too bound IMO ...

Yes, they aren't directly bound, but I don't think it's a problem to have both in the same class. To pass the data to the sidekiq worker we basically only need the `data` and `subscribers` parameters.

And with both things in the same class we can now do something like this:

```ruby
Diaspora::Federation::Dispatcher.build(sender, Retraction.for(target)).dispatch
```

(Even when we don't really need this, but I think that's still nice)

> I can try to make some code changes to show my idea.

If you think you can do something better, go ahead and show me your idea 👍 I don't see how it could help to separate the `data` and `subscriber` yet, I think it gets more complicated. But if you already have an idea, I'm interested.